feat(plugins) optimize redis pipelining

hbagdi opened the pull request:
### Summary

Redis policy for rate-limiting and response rate-limiting plugins uses redis pipelining feature but start and commit multiple pipelines instead of using a single one.

### Full changelog

* Optimize redis storage policy for rate limiting plugin by using a single pipeline
* Optimize redis storage policy for response rate limiting plugin by using a single pipeline

### Issues resolved

Fixes none

thefosk commented:
@hbagdi hello - do you think you would be able to also push a fix to the response-ratelimiting plugin which is _almost_ using the same code?

hbagdi commented:
@thefosk you mean, apply this fix for response-ratelimiting in addition to the current work?

thefosk commented:
@hbagdi yes, it would be very similar but happy to reconsider if that's not the case.

thibaultcha commented on kong/plugins/rate-limiting/policies/init.lua:
```diff
@@ -99,6 +99,7 @@ return {
         end
       end
 
+	  local t = {}
```
Could we come up with a better name for this variable than `t`? Maybe `keys`?

I would also suggest reducing the GC pressure by creating two tables and an index variable here:
```lua
local keys = {}
local expirations = {}
local idx = 0
```

Then, the logic in the `pairs(periods)` loop becomes:
```lua
idx = idx + 1
keys[idx] = cache_key

if not exists or exists == 0 then
  expirations[idx] = EXPIRATIONS[period]
end 
```

Doing this will:
- Only create 2 tables instead of creating a table per limit
- Avoid the O(n) operation implied by the length operator `#` (it is a typical LuaJIT idiom to keep track of indexes separately as we are building lists)
- Reduce branching

Then, the pipelining logic becomes:
```lua
for i = 1, idx do
  red:incrby(keys[i], value)
  if expirations[i] then
    red:expire(keys[i], expirations[i])
  end
end
```

Once again we avoid the `#` operator.

thibaultcha replied:
style: this indentation is also completely off, please respect the 2 spaces indentation encountered across this codebase.

thibaultcha commented on kong/plugins/rate-limiting/policies/init.lua:
```diff
@@ -108,27 +109,35 @@ return {
+      red:init_pipeline()
+      for i = 1,#t,1 do
+        red:incrby(t[i].key, value)
+        ngx.log(ngx.ERR,t[i].key)
```
Please remove those error logging statements as they aren't logging anything particularly useful to the user (and aren't logging an error anyways).

thibaultcha commented on kong/plugins/rate-limiting/policies/init.lua:
```diff
@@ -108,27 +109,35 @@ return {
         return nil, err
       end
-
+  
```
Please remove such undesired changes from your patchset - thanks!

hbagdi commented:
@thefosk, took a quick look on the plugin and adding the optimization there and testing it seems doable. 

@thibaultcha Thanks for your review. I'll address all the comments and update the PR.

hbagdi commented:
@thefosk @thibaultcha 
I've addressed the comments in the recent commits.

Please let me know if there are further changes required here.

Also, I can squash the commits or rename them as you see fit.
